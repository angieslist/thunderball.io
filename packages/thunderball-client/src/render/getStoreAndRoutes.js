import { routerMiddleware, syncHistoryWithStore } from 'react-router-redux';
import configureStore from './configureStore';

/* eslint max-params: 0 */ // this is ok for internal methods
export default function getStoreAndRoutes(
  initialState, createRoutes, historyType, historySyncOptions, injectors = [], pageProps = {}) {
  // Injection point: beforePageMounted
  injectors
    .filter(injector => injector.beforePageMounted)
    .forEach(injector => injector.beforePageMounted(pageProps));

  // Configure store (injectors are called in this function)
  const store = configureStore({
    initialState,
    middleware: [routerMiddleware(historyType)],
  }, injectors, pageProps);

  // Create history by syncing with the redux store
  const history = syncHistoryWithStore(historyType, store, historySyncOptions);

  // Injection point: beforeCreateRoutes
  injectors
    .filter(injector => injector.beforeCreateRoutes)
    .forEach(injector => injector.beforeCreateRoutes(pageProps));

  // Create routes
  const routes = createRoutes(store.getState, store.dispatch);

  // Injection point: afterCreateRoutes
  injectors
    .filter(injector => injector.afterCreateRoutes)
    .forEach(injector => injector.afterCreateRoutes(routes, pageProps));

  return { store, routes, history };
}
