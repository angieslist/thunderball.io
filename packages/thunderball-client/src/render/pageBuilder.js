import React from 'react';
import ReactDOM from 'react-dom';
import { browserHistory } from 'react-router';
import { routerMiddleware, syncHistoryWithStore } from 'react-router-redux';
import Shell from './Shell';
import getStoreAndRoutes from './getStoreAndRoutes';

const render = (Component, props) => {
  ReactDOM.render(
    <Component {...props} />,
    document.getElementById('app'),
  );
};

export default function pageBuilder(createRoutes, injectors = [], defaultLocale, pageProps = {}) {
  // WARNING: This is the client side version of the page rendering,
  // see thunderball/server/browser/render for the server side version.
  // These need to remain somewhat the same.

  // If injectors isn't an array, make it an array with what was passed in
  // This allows for a single injector object
  if (!Array.isArray(injectors)) {
    /* eslint no-param-reassign: 0 */
    injectors = [injectors];
  }

  // Initial state is set isomorphically at first
  const initialState = (typeof window === 'object') ? window.__INITIAL_STATE__ : {};

  const { store, routes, history } = getStoreAndRoutes(
    initialState, createRoutes, browserHistory, { adjustUrlOnReplay: false }, injectors, pageProps);

  const shellProps = { injectors, store, history, pageProps, routes, defaultLocale };

  if (module.hot && typeof module.hot.accept === 'function') {
    module.hot.accept('./Shell', () => {
      const NextShell = require('./Shell');
      render(NextShell, shellProps);
    });
  }

  render(Shell, shellProps);
}
