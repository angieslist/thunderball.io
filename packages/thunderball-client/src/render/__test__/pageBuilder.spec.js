import ReactDOM from 'react-dom';
import getStoreAndRoutes from '../getStoreAndRoutes';
import pageBuilder from '../pageBuilder';

jest.mock('react-dom');
jest.mock('../getStoreAndRoutes');

const mockedStore = Object.freeze({
  dispatch: jest.fn(),
  getState: jest.fn(),
  testStore: true,
});
const mockedRoutes = {
  testRoutes: true,
};
const mockedHistory = {
  testHistory: true,
};
const injector = {
  beforePageMounted: jest.fn(),
  beforeCreateRoutes: jest.fn(),
  afterCreateRoutes: jest.fn(),
};
const pageProps = {
  aPageProp: 'aPagePropValue',
};
const createRoutes = () => mockedRoutes;

getStoreAndRoutes.mockReturnValue({
  store: mockedStore,
  routes: mockedRoutes,
  history: mockedHistory,
});

describe('pageBuilder', () => {
  let app;
  beforeEach(() => {
    app = document.createElement('div');
    app.setAttribute('id', 'app');
    document.getElementsByTagName('body').item(0).appendChild(app);
  });
  it('renders Shell with props into div#app', () => {
    pageBuilder(createRoutes, injector, 'en', pageProps);
    expect(ReactDOM.render.mock.calls).toMatchSnapshot();
  });
});
