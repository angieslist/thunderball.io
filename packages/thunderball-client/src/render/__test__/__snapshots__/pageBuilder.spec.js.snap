// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pageBuilder renders Shell with props into div#app 1`] = `
Array [
  Array [
    <Shell
      defaultLocale="en"
      history={
        Object {
          "testHistory": true,
        }
      }
      injectors={
        Array [
          Object {
            "afterCreateRoutes": [MockFunction],
            "beforeCreateRoutes": [MockFunction],
            "beforePageMounted": [MockFunction],
          },
        ]
      }
      pageProps={
        Object {
          "aPageProp": "aPagePropValue",
        }
      }
      routes={
        Object {
          "testRoutes": true,
        }
      }
      store={
        Object {
          "dispatch": [MockFunction],
          "getState": [MockFunction],
          "testStore": true,
        }
      }
    />,
    <div
      id="app"
    />,
  ],
]
`;
